### PR TITLE
Backport suspendable stream into parity-1.11

### PR DESCRIPTION
**Pull request by seunlanlege**

### CODE REVIEW

File: http/src/lib.rs [hunk 1]
```diff
@@ -438,23 +438,24 @@ fn serve<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>>(
...
 			let allowed_hosts = server_utils::hosts::update(allowed_hosts, &local_addr);
-
+			
 			let http = {
-				let mut http = server::Http::new();
-				http.keep_alive(keep_alive);
-				http.sleep_on_errors(true);
-				http
+				let mut h = server::Http::new();
+				h.keep_alive(keep_alive);
```
Comment (tomusdrw) [inline http/src/lib.rs]: can we get `sleep_on_errors` back? I believe parity-1.11 doesn't have the latest hyper which has this enabled by default.

File: server-utils/src/suspendable_stream.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
+use std::time::{Duration, Instant};
+use tokio_timer::Delay;
+use std::io;
+use core::futures::prelude::*;
+use core::futures::task;
+/// `Incoming` is a stream of incoming sockets
+/// Polling the stream may return a temporary io::Error (for instance if we can't open the connection because of "too many open files" limit)
```
Comment (dvdplm) [inline server-utils/src/suspendable_stream.rs]: Please wrap this line.

File: server-utils/src/suspendable_stream.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+				Ok(item) => {
+					if self.next_delay > self.initial_delay {
+						self.next_delay = self.initial_delay;
+					}
+					return Ok(item)
+				}
+				Err(ref err) => {
+					if connection_error(err) {
+						warn!("Connection Error: {:?}", err);
+						continue
```
Comment (dvdplm) [inline server-utils/src/suspendable_stream.rs]: Now that I look at this again, I wonder if it's really ok to `continue` indefinitely here. If a peer returns e.g. `ConnectionRefused` wouldn't we keep hammering it in an infinite loop?

Comment (seunlanlege) [reply server-utils/src/suspendable_stream.rs]: Actually, this is a server, so `ConnectionError`, would mean there was an error when we tried to accept a connection, that socket could be dropped and we would try to accept a new/different connection.

**Comment by seunlanlege** [comment]
@dvdplm I've added the tests, @tomusdrw looks like `parity-1.11` is still using tokio_core and not the new tokio runtime so i had to use `tokio_timerv0.1` and not `v0.2`

**Comment by seunlanlege** [comment]
I've been building on nightly this whole time :see_no_evil:

**Pull request merged**